File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    void rec(vector<int> vec, vector<int>& nums, int i, int target,vector<vector<int>> &ans){
        if(target==0){
            ans.push_back(vec);
            return;
        }
        if(i==nums.size()){
            return;
        }

        //pick
        if(nums[i]<=target){
            vec.push_back(nums[i]);
            rec(vec,nums,i,target-nums[i],ans);
            vec.pop_back();
        }
        rec(vec,nums,i+1,target,ans);

        return;

    }
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        /*
        [1,2]
        [],nums,i,target
        [],nums,0,3
        pick                not pick
        [1],0,2                 [],1,3
        [1,1],0,1 [1],1,2       [2],2,1
        */
        vector<vector<int>> ans;
        vector<int> vec;
        rec(vec,nums,0,target,ans);
        return ans;
    }
};

File: Data Structures & Algorithms/sum-of-all-subset-xor-totals/submission-0.java
class Solution {
    public int rec(int[] nums,int i,int sum){
        if(i>=nums.length){
            return sum;
        }

        // pick
        int left = rec(nums,i+1,sum ^ nums[i]);

        // not pick
        int right = rec(nums,i+1,sum);

        return left + right;

    }
    public int subsetXORSum(int[] nums) {
        int sum = 0;

        sum = rec(nums,0,sum);


        return sum;
    }
}